From 7fe495e0469bfeb54165371341de93152bd3cd44 Mon Sep 17 00:00:00 2001
From: WordPress Playground contributors <playground@wordpress.org>
Date: Sat, 18 Jul 2026 19:00:00 +0200
Subject: [PATCH 15/15] Use a signature-safe resource destructor on WASI

PHP 7.4's zend_list_free() returns int, while zend_rc_dtor_func_t returns
void. Casting zend_list_free() into the refcounted destructor table happens
to work on conventional native ABIs, but WebAssembly validates call_indirect
signatures and traps when an implicitly released resource reaches that entry.

Keep zend_list_free()'s PHP 7.4 API intact and adapt it through a function with
the exact destructor-table signature.
---
 Zend/zend_variables.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/Zend/zend_variables.c b/Zend/zend_variables.c
index 473b48e278..a91150873a 100644
--- a/Zend/zend_variables.c
+++ b/Zend/zend_variables.c
@@ -33,6 +33,7 @@ static void ZEND_FASTCALL zend_string_destroy(zend_string *str);
 #endif
 static void ZEND_FASTCALL zend_reference_destroy(zend_reference *ref);
 static void ZEND_FASTCALL zend_empty_destroy(zend_reference *ref);
+static void ZEND_FASTCALL zend_resource_destroy(zend_refcounted *p);
 
 typedef void (ZEND_FASTCALL *zend_rc_dtor_func_t)(zend_refcounted *p);
 
@@ -46,7 +47,7 @@ static const zend_rc_dtor_func_t zend_rc_dtor_func[] = {
 	/* IS_STRING       */ (zend_rc_dtor_func_t)zend_string_destroy,
 	/* IS_ARRAY        */ (zend_rc_dtor_func_t)zend_array_destroy,
 	/* IS_OBJECT       */ (zend_rc_dtor_func_t)zend_objects_store_del,
-	/* IS_RESOURCE     */ (zend_rc_dtor_func_t)zend_list_free,
+	/* IS_RESOURCE     */ zend_resource_destroy,
 	/* IS_REFERENCE    */ (zend_rc_dtor_func_t)zend_reference_destroy,
 	/* IS_CONSTANT_AST */ (zend_rc_dtor_func_t)zend_ast_ref_destroy
 };
@@ -57,6 +58,11 @@ ZEND_API void ZEND_FASTCALL rc_dtor_func(zend_refcounted *p)
 	zend_rc_dtor_func[GC_TYPE(p)](p);
 }
 
+static void ZEND_FASTCALL zend_resource_destroy(zend_refcounted *p)
+{
+	(void) zend_list_free((zend_resource *) p);
+}
+
 #if ZEND_DEBUG
 static void ZEND_FASTCALL zend_string_destroy(zend_string *str)
 {
-- 
2.49.0
